src/diagnostics/riscv64/disasm-riscv64.cc

```diff
@@ -378,13 +469,146 @@ int Decoder::FormatFPURegisterOrRoundMode(Instruction* instr,
   UNREACHABLE();
 }
 
+// Handle all C extension register based formatting in this function to reduce the
+// complexity of FormatOption.
+int Decoder::FormatRvcRegister(Instruction* instr, const char* format) {
+  DCHECK_EQ(format[0], 'C');
+  DCHECK(format[1] == 'r' || format[1] == 'f');
+  if (format[2] == 's') {  // 'Crs[12]: Rs register.
+    if (format[3] == '1') {
+      if (format[4] == 's') {   // 'Crs1s: 3-bits register
+        int reg = instr->RvcRs1sValue();
+        if (format[1] == 'r') {
+          PrintRegister(reg);
+        } else if (format[1] == 'f') {
+          PrintFPURegister(reg);
+        }
+        return 5;
+      }
+      int reg = instr->RvcRs1Value();
+      if (format[1] == 'r') {
+        PrintRegister(reg);
+      } else if (format[1] == 'f') {
+        PrintFPURegister(reg);
+      }
+      return 4;
+    } else if (format[3] == '2') {
+      if (format[4] == 's') {   // 'Crs2s: 3-bits register
+        int reg = instr->RvcRs2sValue();
+        if (format[1] == 'r') {
+          PrintRegister(reg);
+        } else if (format[1] == 'f') {
+          PrintFPURegister(reg);
+        }
+        PrintRegister(reg);
+        return 5;
+      }
+      int reg = instr->RvcRs2Value();
+      if (format[1] == 'r') {
+        PrintRegister(reg);
+      } else if (format[1] == 'f') {
+        PrintFPURegister(reg);
+      }
+      return 4;
+    }
+    UNREACHABLE();
+  } else if (format[2] == 'd') {  // 'Crd: rd register.
+    int reg = instr->RvcRdValue();
+    if (format[1] == 'r') {
+      PrintRegister(reg);
+    } else if (format[1] == 'f') {
+      PrintFPURegister(reg);
+    }
+    return 3;
+  }
+  UNREACHABLE();
+}
+
+// Handle all C extension immediates based formatting in this function to reduce the
+// complexity of FormatOption.
+int Decoder::FormatRvcImm(Instruction* instr, const char* format) {
+  DCHECK_EQ(format[0], 'C');
+  DCHECK(STRING_STARTS_WITH(format, "Cimm"));
+  if (format[4] == 'L') {
+    if (format[5] == 'd') {
+      if (format[6] == 's') {
+        DCHECK(STRING_STARTS_WITH(format, "CimmLdsp"));
+        PrintRvcImmLdsp(instr);
+        return 8;
+      }
+      DCHECK(STRING_STARTS_WITH(format, "CimmLd"));
+      PrintRvcImmLd(instr);
+      return 6;
+    } else if (format[5] == 'w') {
+      if (format[6] == 's') {
+        DCHECK(STRING_STARTS_WITH(format, "CimmLwsp"));
+        PrintRvcImmLwsp(instr);
+        return 8;
+      }
+      DCHECK(STRING_STARTS_WITH(format, "CimmLw"));
+      PrintRvcImmLw(instr);
```

Comment:
Add number of immediate bits to the format string for better readability, e.g., `CimmLw` -> `Cimm6Lw`